compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -1017,14 +1037,38 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     #[inline(always)]
     pub(crate) fn const_validate_operand(
         &self,
-        op: &OpTy<'tcx, M::Provenance>,
+        mplace: MPlaceTy<'tcx>,
         path: Vec<PathElem>,
-        ref_tracking: &mut RefTracking<MPlaceTy<'tcx, M::Provenance>, Vec<PathElem>>,
+        ref_tracking: &mut RefTracking<MPlaceTy<'tcx>, Vec<PathElem>>,
         ctfe_mode: CtfeValidationMode,
     ) -> InterpResult<'tcx> {
-        self.validate_operand_internal(op, path, Some(ref_tracking), Some(ctfe_mode))
+        let prov = mplace.ptr().provenance;
+        let path = self.validate_operand_internal(
+            &mplace.into(),
+            path,
+            Some(ref_tracking),
+            Some(ctfe_mode),
+        )?;
+
+        // There was no error, so let's check the rest of the relocations in the pointed-to allocation for
+        // dangling pointers.
```

Comment:
Hm... so at this point we're not just checking the operand, we're checking the entire allocation. This feels a bit unclean but makes sense. Some thoughts:
- The error should indicate that this pointer is somewhere in this here, likely inside a union or otherwise in the gaps between the fields. In fact we could have a similar check on `visit_union`, couldn't we, so that here we know it has to be some odd padding situation?
- Can't mutable pointers be checked the same way? Then interning doesn't have to return anything.